# Library

In [362]:
import pandas as pd
import os
from datetime import datetime
import numpy as np
from tqdm import tqdm
from glob import glob

# fold path

In [363]:
# # backup
# fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis'
# norm_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis/norm'
# total_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis/norm/total'
# error_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis/error'
# total_error_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis/error/total'
# weird_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis/weird'
# nan_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis/nan'
# nad_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/emis/nan/nad'

In [364]:
# update2
fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2'
norm_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/norm'
total_norm_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/norm/total'
error_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/error'
total_error_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/error/total'
weird_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/weird'
total_weird_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/weird/total'
nan_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/nan'
total_nan_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/nan/total'
submit_fold = 'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/submit'

# File Load
- 정상 배인번호 파일
- 비정상 배인번호 파일
- 제작사 이상 파일

## 정상 배인번호 파일

In [365]:
# about 41s
norm = pd.DataFrame()
for addr in tqdm(glob(os.path.join(norm_fold, '*.csv'))):
    temp = pd.read_csv(addr, encoding='cp949', low_memory=False)
    norm = pd.concat([norm, temp], ignore_index=True)

100%|██████████| 134/134 [00:41<00:00,  3.25it/s]


In [366]:
norm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1140492 entries, 0 to 1140491
Data columns (total 69 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   차대번호       1140492 non-null  object 
 1   차량번호_x     1140492 non-null  object 
 2   제원관리번호     1140492 non-null  object 
 3   차종_원본      1140492 non-null  object 
 4   용도         1140492 non-null  object 
 5   최초등록일      1140492 non-null  int64  
 6   연식         1140492 non-null  int64  
 7   제작일자       1140491 non-null  float64
 8   검사유효일      1140490 non-null  float64
 9   배출가스인증번호   1140492 non-null  object 
 10  등급         1140492 non-null  int64  
 11  차명_x       1140492 non-null  object 
 12  차종분류       1140492 non-null  object 
 13  차종유형       1140492 non-null  object 
 14  자동차형식      1140492 non-null  object 
 15  제작사명       1140492 non-null  object 
 16  연료         1140492 non-null  object 
 17  엔진형식_x     1140492 non-null  object 
 18  총중량        1140492 non-null  float64
 19  

## 비정상 배인번호 파일

In [369]:
error = pd.DataFrame()
for addr in tqdm(glob(os.path.join(error_fold, '*.csv'))):
    temp = pd.read_csv(addr, encoding='cp949', low_memory=False)
    error = pd.concat([error, temp], ignore_index=True)

100%|██████████| 71/71 [00:00<00:00, 80.04it/s]


In [370]:
error.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 69 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   차대번호       1309 non-null   object 
 1   차량번호_x     1309 non-null   object 
 2   제원관리번호     1309 non-null   object 
 3   차종_원본      1309 non-null   object 
 4   용도         1309 non-null   object 
 5   최초등록일      1309 non-null   int64  
 6   연식         1309 non-null   int64  
 7   제작일자       1309 non-null   float64
 8   검사유효일      1309 non-null   float64
 9   배출가스인증번호   1309 non-null   object 
 10  등급         1309 non-null   int64  
 11  차명_x       1309 non-null   object 
 12  차종분류       1309 non-null   object 
 13  차종유형       1309 non-null   object 
 14  자동차형식      1309 non-null   object 
 15  제작사명       1304 non-null   object 
 16  연료         1309 non-null   object 
 17  엔진형식_x     1309 non-null   object 
 18  총중량        1309 non-null   float64
 19  적재중량       1309 non-null   float64
 20  엔진출력    

## 제작사 이상 파일

In [371]:
glob(os.path.join(weird_fold, '*.csv'))

['D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2/weird\\weird.csv']

In [372]:
weird = pd.DataFrame()
for addr in tqdm(glob(os.path.join(weird_fold, '*.csv'))):
    temp = pd.read_csv(addr, encoding='cp949')
    weird = pd.concat([weird, temp], ignore_index=True)

100%|██████████| 1/1 [00:00<00:00, 13.51it/s]


In [373]:
weird.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4862 entries, 0 to 4861
Data columns (total 69 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   차대번호       4862 non-null   object 
 1   차량번호_x     4862 non-null   object 
 2   제원관리번호     4862 non-null   object 
 3   차종_원본      4862 non-null   object 
 4   용도         4862 non-null   object 
 5   최초등록일      4862 non-null   int64  
 6   연식         4862 non-null   int64  
 7   제작일자       4862 non-null   float64
 8   검사유효일      4862 non-null   float64
 9   배출가스인증번호   4862 non-null   object 
 10  등급         4862 non-null   int64  
 11  차명_x       4862 non-null   object 
 12  차종분류       4862 non-null   object 
 13  차종유형       4862 non-null   object 
 14  자동차형식      4862 non-null   object 
 15  제작사명       4850 non-null   object 
 16  연료         4862 non-null   object 
 17  엔진형식_x     4862 non-null   object 
 18  총중량        4862 non-null   float64
 19  적재중량       4857 non-null   float64
 20  엔진출력    

In [374]:
weird['배인번호_수정'].isnull().sum()

0

## 미부여

In [375]:
nad = pd.DataFrame()
for addr in tqdm(glob(os.path.join(nan_fold, '*.csv'))):
    temp = pd.read_csv(addr, encoding='cp949', low_memory=False)
    nad = pd.concat([nad, temp], ignore_index=True)

100%|██████████| 1/1 [00:00<00:00, 10.97it/s]


In [376]:
nad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7150 entries, 0 to 7149
Data columns (total 69 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   차대번호       7150 non-null   object 
 1   차량번호_x     7150 non-null   object 
 2   제원관리번호     7150 non-null   object 
 3   차종_원본      7150 non-null   object 
 4   용도         7150 non-null   object 
 5   최초등록일      7150 non-null   int64  
 6   연식         7150 non-null   int64  
 7   제작일자       7147 non-null   float64
 8   검사유효일      7149 non-null   float64
 9   배출가스인증번호   0 non-null      float64
 10  등급         7150 non-null   int64  
 11  차명_x       7150 non-null   object 
 12  차종분류       7150 non-null   object 
 13  차종유형       7150 non-null   object 
 14  자동차형식      7149 non-null   object 
 15  제작사명       7149 non-null   object 
 16  연료         7150 non-null   object 
 17  엔진형식_x     7150 non-null   object 
 18  총중량        7150 non-null   float64
 19  적재중량       7129 non-null   float64
 20  엔진출력    

In [377]:
norm.shape, error.shape, weird.shape, nad.shape

((1140492, 69), (1309, 69), (4862, 69), (7150, 69))

In [378]:
# 1153813
norm.shape[0] + error.shape[0] + weird.shape[0] + nad.shape[0]

1153813

## 미기입 확인

In [379]:
check_nan_car = pd.read_excel(os.path.join(fold, '[BD1] 8. 배출가스 인증번호 미기입 차량(23.03.16)_2.xlsx'))
check_nan_car.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7150 entries, 0 to 7149
Data columns (total 30 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   차대번호      7150 non-null   object 
 1   차량번호      7150 non-null   object 
 2   제원관리번호    7150 non-null   object 
 3   차종        7150 non-null   object 
 4   용도        7150 non-null   object 
 5   최초등록일     7150 non-null   int64  
 6   연식        7150 non-null   int64  
 7   제작일자      7147 non-null   float64
 8   검사유효일     7149 non-null   float64
 9   배출가스인증번호  0 non-null      float64
 10  배인번호_수정   6525 non-null   object 
 11  등급        7150 non-null   int64  
 12  등급_수정     7150 non-null   int64  
 13  차명        7150 non-null   object 
 14  차종분류      7150 non-null   object 
 15  차종유형      7150 non-null   object 
 16  자동차형식     7149 non-null   object 
 17  제작사명      7149 non-null   object 
 18  연료        7150 non-null   object 
 19  엔진형식      7150 non-null   object 
 20  총중량       7150 non-null   int6

# 정제

In [380]:
norm['등급_수정'] = norm['등급_수정'].astype('str')

## 병합 파일 출력

In [381]:
select_col = ['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']

### 정상 파일

In [382]:
concat_norm = norm[select_col]
concat_norm = concat_norm.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
concat_norm.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [383]:
# about 33s
concat_norm.to_csv(os.path.join(total_norm_fold, 'concat_norm.csv'), encoding='cp949', index=False)

### 비정상 파일

In [384]:
concat_error = error[select_col]
concat_error = concat_error.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
concat_error.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [385]:
concat_error.to_csv(os.path.join(total_error_fold, 'concat_error.csv'), index=False, encoding='cp949')

### 제작사 이상 파일

In [386]:
concat_weird = weird[select_col]
concat_weird = concat_weird.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
concat_weird.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [387]:
concat_weird.to_csv(os.path.join(total_weird_fold, 'concat_weird.csv'), index=False, encoding='cp949')

### 미부여 파일

In [388]:
concat_nad = nad[select_col]
concat_nad = concat_nad.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
concat_nad.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [389]:
concat_nad.to_csv(os.path.join(total_nan_fold, 'concat_nad.csv'), index=False, encoding='cp949')

## 미기입 확인
- 필요한 부분만 추출

In [390]:
sliced_check_nan_car = check_nan_car.loc[check_nan_car['배인번호_수정'].notnull() == True].reset_index(drop=True)
sliced_check_nan_car.shape

(6525, 30)

# 매칭

## 컬럼명 설정

In [391]:
# 컬럼명 설정
col_car_name = '차명_x'
col_engine_form = '엔진형식_x'

## 미부여

### 1\. 유로4인증대장 매칭

In [392]:
# 유로4 / result파일
# 제작사명 : 크라이슬러 / 에프씨에이코리아(주)
# 차명으로 매칭
# {'300C' : '6MY-DK-14-1', '짚 커맨더' : '6MY-DK-14-2', '니트로' : '7MY-DK-14-4', '짚 니트로' : '7MY-DK-14-4', '짚 랭글러' : '7MY-DK-14-5', '짚그랜드체로키' : '8MY-DK-14-01, '짚 그랜드 체로키' : '8MY-DK-14-01', '그랜드 보이저' : '8MY-DK-14-02', '세브링' : '8MY-DK-14-06'}

In [393]:
# 유로4 / result파일
# 제작사명 : 볼보 / (주)볼보자동차코리아
# 차명으로 매칭
# {'볼보 S60D5' : '5MY-PA-14-17', '볼보 S80D5' : '6MY-PA-14-4', '볼보 XC70D5' : '8MY-PA-14-03', '볼보 XC60D5' : '9MY-PA-13-01'}

In [394]:
# 유로4 / result파일
# 제작사명 : 랜드로버 / (주)재규어랜드로버코리아
# 차명으로 매칭
# {'레인지로버 스포츠 TDV6' : '7MY-PA-14-2'}

In [395]:
# 유로4 / result파일
# 제작사명 : 재규어 / (주)재규어랜드로버코리아
# 차명으로 매칭
# {'재규어 XJ 2.7D' : '7MY-PA-14-7', '재규어 X-TYPE 2.2D' : '8MY-PA-14-04', '재규어 XF 2.7D' : '8MY-PA-14-05'}

In [396]:
# 유로4 / result파일
# 제작사명 : 재규어 / (주)볼보자동차코리아
# 차명으로 매칭
# {'재규어 XJ 2.7D' : '7MY-PA-14-7', '재규어 X-TYPE 2.2D' : '8MY-PA-14-04', '재규어 XF 2.7D' : '8MY-PA-14-05'}

In [397]:
# 유로4 / result파일
# 제작사명 : 포드 / 포드세일즈서비스코리아(유)
# 차명으로 매칭
# {'에스맥스' : '8MY-FK-14-1', '몬데오' : '8MY-FK-14-2'}

In [398]:
# 유로4 / result파일
# 제작사명 : 푸조 / 한불모터스(주)
# 배기량
# {1997 : '5MY-HB-14-1', 2721 : '5MY-HB-14-4', 1560 : '8MY-HB-14-1'}

In [399]:
# 2129, 5021 (y, n)
productor = '에프씨에이코리아(주)'
dict1 = {'300C' : '6MY-DK-14-1', '짚 커맨더' : '6MY-DK-14-2', '니트로' : '7MY-DK-14-4', '짚 니트로' : '7MY-DK-14-4', '짚 랭글러' : '7MY-DK-14-5', 
         '짚그랜드체로키' : '8MY-DK-14-01', '짚 그랜드 체로키' : '8MY-DK-14-01', '그랜드 보이저' : '8MY-DK-14-02', '세브링' : '8MY-DK-14-06'}
for k, v in dict1.items():
    nad.loc[(nad['제작사명'] == productor) & (nad[col_car_name] == k), '배인번호_수정'] = v

In [401]:
# 3265, 3885
productor = '(주)볼보자동차코리아'
dict2 = {'볼보 S60D5' : '5MY-PA-14-17', '볼보 S80D5' : '6MY-PA-14-4', '볼보 XC70D5' : '8MY-PA-14-03', '볼보 XC60D5' : '9MY-PA-13-01'}
for k, v in dict2.items():
    nad.loc[(nad['제작사명'] == productor) & (nad[col_car_name] == k), '배인번호_수정'] = v

In [403]:
# 3268, 3882
productor = '(주)재규어랜드로버코리아'
dict3 = {'레인지로버 스포츠 TDV6' : '7MY-PA-14-2'}
for k, v in dict3.items():
    nad.loc[(nad['제작사명'] == productor) & (nad[col_car_name] == k), '배인번호_수정'] = v

In [405]:
# 3539, 3611
productor = '(주)재규어랜드로버코리아'
dict4 = {'재규어 XJ 2.7D' : '7MY-PA-14-7', '재규어 X-TYPE 2.2D' : '8MY-PA-14-04', '재규어 XF 2.7D' : '8MY-PA-14-05'}
for k, v in dict4.items():
    nad.loc[(nad['제작사명'] == productor) & (nad[col_car_name] == k), '배인번호_수정'] = v

In [407]:
# 3637, 3513
productor = '(주)볼보자동차코리아'
dict5 = {'재규어 XJ 2.7D' : '7MY-PA-14-7', '재규어 X-TYPE 2.2D' : '8MY-PA-14-04', '재규어 XF 2.7D' : '8MY-PA-14-05'}
for k, v in dict5.items():
    nad.loc[(nad['제작사명'] == productor) & (nad[col_car_name] == k), '배인번호_수정'] = v

In [409]:
# 3736, 3414
productor = '포드세일즈서비스코리아(유)'
dict6 = {'에스맥스' : '8MY-FK-14-1', '몬데오' : '8MY-FK-14-2'}
for k, v in dict6.items():
    nad.loc[(nad['제작사명'] == productor) & (nad[col_car_name] == k), '배인번호_수정'] = v

In [411]:
# 6294, 856
productor = '한불모터스(주)'
dict7 = {1997 : '5MY-HB-14-1', 2721 : '5MY-HB-14-4', 1560 : '8MY-HB-14-1'}
for k, v in dict7.items():
    nad.loc[(nad['제작사명'] == productor) & (nad['배기량'] == k), '배인번호_수정'] = v

In [412]:
nad1y = nad.loc[nad['배인번호_수정'].isnull() == False].reset_index(drop=True)
nad1n = nad.loc[nad['배인번호_수정'].isnull() == True].reset_index(drop=True)
nad1y.shape, nad1n.shape

((6294, 69), (856, 69))

#### [출력] 처리된 미기입 차량1

In [413]:
submit_nad1y = nad1y[select_col]
submit_nad1y = submit_nad1y.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_nad1y.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [414]:
submit_nad1n = nad1n[select_col]
submit_nad1n = submit_nad1n.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_nad1n.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [415]:
submit_nad1y.to_csv(os.path.join(total_nan_fold, 'nad1y.csv'), index=False, encoding='cp949')
submit_nad1n.to_csv(os.path.join(total_nan_fold, 'nad1n.csv'), index=False, encoding='cp949')

### 2\. 정상 데이터와 비교1
- 동일 제번, 자동차형식, 엔진형식으로 매칭 : 5개 매칭
- 동일 자동차형식, 엔진형식으로 매칭 : 139개 매칭

In [416]:
norm.drop_duplicates(['배출가스인증번호', '제원관리번호', '자동차형식', col_engine_form]).shape

(4547, 69)

In [417]:
match_data = norm.drop_duplicates(['배출가스인증번호', '제원관리번호', '자동차형식', col_engine_form])[['배출가스인증번호', '제원관리번호', '자동차형식', col_engine_form]].values
match_data[:3]

array([['4MY-AD-20', 2420000100031305, '4F', 'BMK'],
       ['4MY-DC-33', 'A0310006701423306', 'K9ARF5', 'DV11'],
       ['4MY-PA-06', 920000300131205, '2.0LN5DR', '204D']], dtype=object)

In [418]:
for emis, no, car, engine in tqdm(match_data):
    # nad1n.loc[(nad1n['제원관리번호'] == no) & (nad1n['자동차형식'] == car) & (nad1n['엔진형식_x'] == engine), '배출가스인증번호'] = emis
    nad1n.loc[(nad1n['자동차형식'] == str(car)) & (nad1n[col_engine_form] == str(engine)), '배인번호_수정'] = emis

100%|██████████| 4547/4547 [00:02<00:00, 1752.55it/s]


In [419]:
# (y, n) = (139, 717)
nad2y = nad1n.loc[nad1n['배인번호_수정'].isnull() == False].reset_index(drop=True)
nad2n = nad1n.loc[nad1n['배인번호_수정'].isnull() == True].reset_index(drop=True)
nad2y.shape, nad2n.shape

((139, 69), (717, 69))

In [420]:
nad2y['제작사명'].unique()

array(['(주)볼보자동차코리아', '비엠더블유코리아(주)', '자기인증면제차량', '볼보트럭코리아(주)',
       '르노삼성자동차(주)', '현대자동차(주)', '한국지엠주식회사', '기아자동차(주)'], dtype=object)

#### [출력] 처리된 미기입 차량2

In [421]:
submit_nad2y = nad2y[select_col]
submit_nad2y = submit_nad2y.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_nad2y.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [422]:
submit_nad2n = nad2n[select_col]
submit_nad2n = submit_nad2n.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_nad2n.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [423]:
submit_nad2y.to_csv(os.path.join(total_nan_fold, 'nad2y.csv'), index=False, encoding='cp949')
submit_nad2n.to_csv(os.path.join(total_nan_fold, 'nad2n.csv'), index=False, encoding='cp949')

### 3\. 정상 데이터와 비교2
- 차대번호 11자리까지 동일하면 동일한 등급으로 선정

In [424]:
nad2n.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

In [425]:
norm['차대번호'] = norm['차대번호'].astype('str')
nad2n['차대번호'] = nad2n['차대번호'].astype('str')

In [426]:
norm_vin11 = norm['차대번호'].str[:11].to_list()

In [427]:
norm_vin11[:5]

['WAUZZZ4F16N', 'WAUZZZ4F16N', 'WAUZZZ4F86N', 'WAUZZZ4F36N', 'WAUZZZ4F26N']

In [428]:
nad2n_vin11 = nad2n['차대번호'].str[:11].to_list()
nad2n_vin11[:5]

['KL3M8BNF37K', 'WJMS2NTH484', 'WJMS2NTH484', 'WJMS2NSK074', 'WJMS2NUH484']

In [429]:
idx3 = []
for i, one in tqdm(enumerate(nad2n_vin11)):
    if one in norm_vin11:
        idx3.append(i)
len(idx3)

717it [00:10, 66.18it/s]


7

In [430]:
nad2n.loc[idx3, ['차대번호', '제원관리번호', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '제작사명', col_car_name, '자동차형식', col_engine_form]]

차대번호             제원관리번호  배출가스인증번호 배인번호_수정  등급  등급_수정  \
6    KMFLA18KP8C027362  B3C10016600003307       NaN     NaN   4      4   
399  KNCWF03429K377309  B3510001600014108       NaN     NaN   4      4   
436  KNCSHZ74CCK639710  B1T10003500013111       NaN     NaN   4      4   
613  KMFLA18KP8C026863  B9G10003300003307       NaN     NaN   4      4   
620  KNCSHZ74CCK639187  B1T10003500013111       NaN     NaN   4      4   
715  KMFLA18KP8C031274  B1R10005100003308       NaN     NaN   4      4   
716  KMCLK18KP8C035195  A0810005209353308       NaN     NaN   4      4   

             제작사명             차명_x              자동차형식 엔진형식_x  
6       한국특장기술(주)         한국특장기술트럭  K4.5C2-7TD-H4.5EU   D6GA  
399     (주)동해기계항공  동해12M고소작업차-0801     DHS1200HM-0801     J3  
436     주식회사 한서정공   한서4륜구동다목적방역방제차          HS-1500KF     J3  
613         정부쓰리축      정부5톤플러스카고트럭           JBH-5LLP   D6GA  
620     주식회사 한서정공   한서4륜구동다목적방역방제차          HS-1500KF     J3  
715  신정개발특장차 주식회사     신정4㎥급진공노면총소차           RS07GMS2   D6GA  
716      현대자동차(주)        메가트럭군물자수송   HD050-8MT2-LGc-A   D6GA

In [431]:
norm_vin11.index('KMCLK18KP8C')

764822

In [432]:
norm.loc[764822, ['차대번호', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정']]

차대번호        KMCLK18KP8C031162
배출가스인증번호         7MY-HD-24-30
배인번호_수정          7MY-HD-24-30
등급                          4
등급_수정                       4
Name: 764822, dtype: object

In [433]:
idx3_1 = []
for i, one in tqdm(enumerate(norm_vin11)):
    if one in nad2n_vin11:
        idx3_1.append(i)
len(idx3_1)

1140492it [00:10, 113025.10it/s]


6425

### 4\. 미기입 확인 적용

In [434]:
nad2n.shape

(717, 69)

In [435]:
nad2n.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

In [436]:
sliced_check_nan_car.shape

(6525, 30)

In [437]:
pick_car = sliced_check_nan_car.drop_duplicates(['배인번호_수정', '등급_수정', '차명', '자동차형식', '엔진형식'])
pick_car.shape

(91, 30)

In [438]:
for m_emis, m_grade, car_name, car_form, engine_form in tqdm(pick_car[['배인번호_수정', '등급_수정', '차명', '자동차형식', '엔진형식']].values):
    nad2n.loc[(nad2n[col_car_name] == car_name) & (nad2n['자동차형식'] == car_form) & (nad2n[col_engine_form] == engine_form), ['배인번호_수정', '등급_수정']] = m_emis, m_grade

100%|██████████| 91/91 [00:00<00:00, 550.68it/s]


In [439]:
# (y, n) = (87, 630)
nad3y = nad2n.loc[nad2n['배인번호_수정'].isnull() == False].reset_index(drop=True)
nad3n = nad2n.loc[nad2n['배인번호_수정'].isnull() == True].reset_index(drop=True)
nad3y.shape, nad3n.shape

((87, 69), (630, 69))

#### [출력] 처리된 미기입 차량3

In [440]:
submit_nad3y = nad3y[select_col]
submit_nad3y = submit_nad3y.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_nad3y.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [441]:
submit_nad3n = nad3n[select_col]
submit_nad3n = submit_nad3n.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_nad3n.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [442]:
submit_nad3y.to_csv(os.path.join(total_nan_fold, 'nad3y.csv'), index=False, encoding='cp949')
submit_nad3n.to_csv(os.path.join(total_nan_fold, 'nad3n.csv'), index=False, encoding='cp949')

## 제작사 이상
- 4862 -> 4715, 147 개 처리 

### [출력] weird 파일

In [443]:
weird.shape

(4862, 69)

In [444]:
weird.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

### 1\. 유로4인증대장과 매칭

In [445]:
weird['제작사명'].unique()

array(['현대자동차(주)', '창림모아츠(주)', '(주)에이엠티코리아', '주식회사 평강비아이엠(B.I.M)',
       '(주)지엠케이', '드림특장(주)', '(주)신광테크놀러지', '제이에스이엔지', '(주)용진특장', '(주)엠뱅크',
       '수성특장(주)', '(주)세정모터스', '(주)엘위포투', nan, '(주)성진테크', '(주)밴텍캠핑카',
       '자기인증면제차량', '르노삼성자동차(주)', '(주) 삼오특장자동차', '(주) 골드밴', '(주)애드터치',
       '기아자동차(주)', '해양산업', '씨엔라인', '(주)동해기계항공', '(주) 오텍', '(주)호룡',
       '(주)사자후', '(주)무빙애드', '한아에스에스(주)', '주식회사 천하', '(주)드림티엔에스',
       '반도특장(주)', '라헬(주)', '동아특장공업사', '녹산대진특장차', '(주) 고호산업', '에스알(주)',
       '주식회사 카셈(KASEM Co.,Ltd)', '(주)다산중공업', '(주)에코카', '주식회사 유아이테크',
       '세화자동차', '(주)마노', '(주)맥스리무진', '고려에스티엠(주)', '카올넷', '에프씨에이코리아(주)',
       '쌍용자동차(주)', '비엠더블유코리아(주)', '한불모터스(주)', '(주)팔도', '(주)한국토미',
       '(주)포멕특장', '선진정공(주)', '(주)한국쓰리축', '한국특장기술(주)', '디앨 주식회사',
       '씨제이대한통운(주)', '(주)한국차체', '건민특장 주식회사', '(주)화이어텍', '한국쓰리축공업(주)',
       '정우정공 (주)', '우림특장산업', '탱크코리아(주)', '정부쓰리축', '부커셰링코리아 (주)',
       '신정개발특장차 주식회사', '엠에스특장공업(주)', '(주)화성특장', '(주)한국메리트특장',
       '(주)케이지비고려골든박스', '한빛테크원(주)', '(주)뉴텍티앤엠',

In [446]:
# 한불모터스(주)
# 비엠더블유코리아(주)
# 르노삼성자동차(주)
# 에프씨에이코리아(주)
# 기아자동차(주)
# 쌍용자동차(주)
# 현대자동차(주)

In [447]:
# 유로4 / result파일
# 제작사명 : 푸조 / 한불모터스(주)
# 배기량
# {1997 : '5MY-HB-14-1', 2721 : '5MY-HB-14-4', 1560 : '8MY-HB-14-1'}

In [448]:
# 유로4 / result파일
# 제작사명 : 크라이슬러 / 에프씨에이코리아(주)
# 차명으로 매칭
# {'300C' : '6MY-DK-14-1', '짚 커맨더' : '6MY-DK-14-2', '니트로' : '7MY-DK-14-4', '짚 니트로' : '7MY-DK-14-4', '짚 랭글러' : '7MY-DK-14-5', '짚그랜드체로키' : '8MY-DK-14-01, '짚 그랜드 체로키' : '8MY-DK-14-01', '그랜드 보이저' : '8MY-DK-14-02', '세브링' : '8MY-DK-14-06'}

In [449]:
weird['배인번호_수정_확인'] = np.nan

In [450]:
# 8, 4854(y, n)
productor = '한불모터스(주)'
dict1 = {1997 : '5MY-HB-14-1', 2721 : '5MY-HB-14-4', 1560 : '8MY-HB-14-1'}
for k, v in dict1.items():
    weird.loc[(weird['제작사명'] == productor) & (weird['배기량'] == k), ['배인번호_수정', '배인번호_수정_확인']] = v, v

In [452]:
# 132, 4730
productor = '에프씨에이코리아(주)'
dict2 = {'300C' : '6MY-DK-14-1', '짚 커맨더' : '6MY-DK-14-2', '니트로' : '7MY-DK-14-4', '짚 니트로' : '7MY-DK-14-4', '짚 랭글러' : '7MY-DK-14-5', 
         '짚그랜드체로키' : '8MY-DK-14-01', '짚 그랜드 체로키' : '8MY-DK-14-01', '그랜드 보이저' : '8MY-DK-14-02', '세브링' : '8MY-DK-14-06'}
for k, v in dict2.items():
    weird.loc[(weird['제작사명'] == productor) & (weird[col_car_name] == k), ['배인번호_수정', '배인번호_수정_확인']] = v, v

In [453]:
weird1y = weird.loc[weird['배인번호_수정_확인'].isnull() == False].reset_index(drop=True)
weird1n = weird.loc[weird['배인번호_수정_확인'].isnull() == True].reset_index(drop=True)
weird1y.shape, weird1n.shape

((132, 70), (4730, 70))

In [454]:
weird.loc[weird['제작사명'] == '한불모터스(주)', ['배출가스인증번호', '배인번호_수정', '제작사명']].shape

(8, 3)

In [455]:
weird.loc[weird['제작사명'] == '에프씨에이코리아(주)', ['배출가스인증번호', '배인번호_수정', '제작사명', col_car_name]].shape

(127, 4)

In [456]:
weird.loc[(weird['제작사명'] == '에프씨에이코리아(주)') & (weird[col_car_name] == '짚 랭글러'), ['배출가스인증번호', '배인번호_수정', '제작사명', col_car_name]].shape

(124, 4)

In [457]:
weird['배인번호_수정_확인'].isnull().sum()

4730

#### [출력] 처리된 제작사 이상 차량1

In [458]:
submit_weird1y = weird1y[select_col]
submit_weird1y = submit_weird1y.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_weird1y.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [459]:
submit_weird1n = weird1n[select_col]
submit_weird1n = submit_weird1n.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_weird1n.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [460]:
submit_weird1y.to_csv(os.path.join(total_weird_fold, 'weird1y.csv'), index=False, encoding='cp949')
submit_weird1n.to_csv(os.path.join(total_weird_fold, 'weird1n.csv'), index=False, encoding='cp949')

### 2\. 정상 데이터와 비교
- 총 4730
- 자형, 엔형 매칭 : 14

In [461]:
for emis, no, car, engine in tqdm(match_data):
    # weird1n.loc[(weird1n['제원관리번호'] == no) & (weird1n['자동차형식'] == car) & (weird1n['엔진형식_x'] == engine), '배인번호_수정'] = emis
    weird1n.loc[(weird1n['자동차형식'] == car) & (weird1n[col_engine_form] == engine), ['배인번호_수정', '배인번호_수정_확인']] = emis, emis

100%|██████████| 4547/4547 [00:07<00:00, 609.91it/s]


In [462]:
weird2y = weird1n.loc[weird1n['배인번호_수정_확인'].isnull() == False].reset_index(drop=True)
weird2n = weird1n.loc[weird1n['배인번호_수정_확인'].isnull() == True].reset_index(drop=True)
weird2y.shape, weird2n.shape

((14, 70), (4716, 70))

In [469]:
weird2n['배인번호_수정_확인'].isnull().sum()

4716

#### [출력] 처리된 제작사 이상 차량2

In [463]:
submit_weird2y = weird2y[select_col]
submit_weird2y = submit_weird2y.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_weird2y.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [464]:
submit_weird2n = weird2n[select_col]
submit_weird2n = submit_weird2n.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식', '차종_원본':'차종'})
submit_weird2n.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [465]:
submit_weird2y.to_csv(os.path.join(total_weird_fold, 'weird2y.csv'), index=False, encoding='cp949')
submit_weird2n.to_csv(os.path.join(total_weird_fold, 'weird2n.csv'), index=False, encoding='cp949')

### 3\. 미기입 확인 적용

In [466]:
pick_car.columns

Index(['차대번호', '차량번호', '제원관리번호', '차종', '용도', '최초등록일', '연식', '제작일자', '검사유효일',
       '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명', '차종분류', '차종유형', '자동차형식',
       '제작사명', '연료', '엔진형식', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명',
       '시군구명', '소유자구분', '말소여부', '일치유무'],
      dtype='object')

In [467]:
for m_emis, m_grade, car_name, car_form, engine_form in tqdm(pick_car[['배인번호_수정', '등급_수정', '차명', '자동차형식', '엔진형식']].values):
    weird2n.loc[(weird2n[col_car_name] == car_name) & (weird2n['자동차형식'] == car_form) & (weird2n[col_engine_form] == engine_form), ['배인번호_수정', '배인번호_수정_확인', '등급_수정']] = m_emis, m_emis, m_grade

100%|██████████| 91/91 [00:00<00:00, 444.39it/s]


In [468]:
# (y, n) = (0, 4716)
weird3y = weird2n.loc[weird2n['배인번호_수정_확인'].notnull() == True].reset_index(drop=True)
weird3n = weird2n.loc[weird2n['배인번호_수정_확인'].isnull() == True].reset_index(drop=True)
weird3y.shape, weird3n.shape

((0, 70), (4716, 70))

### 남은 차량 중 메이저 제작사 확인
- 에프씨에이코리아(주), 기아자동차(주)

In [470]:
weird2n['제작사명'].unique()

array(['창림모아츠(주)', '(주)에이엠티코리아', '주식회사 평강비아이엠(B.I.M)', '(주)지엠케이',
       '드림특장(주)', '(주)신광테크놀러지', '제이에스이엔지', '(주)용진특장', '(주)엠뱅크', '수성특장(주)',
       '(주)세정모터스', '(주)엘위포투', nan, '(주)성진테크', '(주)밴텍캠핑카', '(주) 삼오특장자동차',
       '(주) 골드밴', '(주)애드터치', '기아자동차(주)', '해양산업', '씨엔라인', '(주)동해기계항공',
       '(주) 오텍', '(주)호룡', '(주)사자후', '(주)무빙애드', '한아에스에스(주)', '주식회사 천하',
       '(주)드림티엔에스', '반도특장(주)', '라헬(주)', '동아특장공업사', '녹산대진특장차', '(주) 고호산업',
       '에스알(주)', '주식회사 카셈(KASEM Co.,Ltd)', '(주)다산중공업', '(주)에코카',
       '주식회사 유아이테크', '세화자동차', '(주)마노', '고려에스티엠(주)', '카올넷', '쌍용자동차(주)',
       '자기인증면제차량', '(주)팔도', '(주)한국토미', '(주)포멕특장', '선진정공(주)', '(주)한국쓰리축',
       '한국특장기술(주)', '디앨 주식회사', '씨제이대한통운(주)', '(주)한국차체', '건민특장 주식회사',
       '(주)화이어텍', '한국쓰리축공업(주)', '정우정공 (주)', '우림특장산업', '탱크코리아(주)', '정부쓰리축',
       '부커셰링코리아 (주)', '신정개발특장차 주식회사', '엠에스특장공업(주)', '(주)화성특장',
       '(주)한국메리트특장', '(주)케이지비고려골든박스', '한빛테크원(주)', '(주)뉴텍티앤엠', '(주)광림',
       '(주)함코', '동양자동차공업(주)', '(주)수산씨에스엠', '(주)우리특장', '(주)에버다임',
       '(주)대진정공', '

In [471]:
weird2n.loc[weird2n['제작사명'] == '에프씨에이코리아(주)', ['배출가스인증번호', '배인번호_수정', '제원관리번호', '자동차형식', col_engine_form, col_car_name]]

배출가스인증번호       배인번호_수정             제원관리번호   자동차형식   엔진형식_x    차명_x
4711  7MY-DK-24-04  7MY-DK-24-04  00220003500014305  1844LS  OM501LA  Actros
4712  7MY-DK-24-04  7MY-DK-24-04  00220003500014305  1844LS  OM501LA  Actros
4713  7MY-DK-24-04  7MY-DK-24-04  00220003500014305  1844LS  OM501LA  Actros

In [472]:
weird2n.loc[weird2n['제작사명'] == '기아자동차(주)', ['배출가스인증번호', '배인번호_수정', '제원관리번호', '자동차형식', col_engine_form, col_car_name]]

배출가스인증번호       배인번호_수정             제원관리번호          자동차형식 엔진형식_x  \
218  7MY-HD-14-68  7MY-HD-14-68  A0110003121643212  PUL6BD-L9-LFT   D4CB   

          차명_x  
218  봉고Ⅲ 활어수송차

# [병합] 처리한 데이터 병합
- norm, error, nad1y, nad2y, nad2n, weird1y, weird2y, weird2n

- 파일 처리 과정
    - nad
        - nad1y
        - nad1n
            - nad2y
            - nad2n
                - nad3y
                - nad3n
    - weird
        - weird1y
        - weird1n
            - weird2y
            - weird2n

In [473]:
df = pd.concat([norm, error, nad1y, nad2y, nad3y, nad3n, weird1y, weird2y, weird2n], ignore_index=True)
df.shape

(1153813, 70)

In [474]:
df.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고', '배인번호_수정_확인'],
      dtype='object')

## [출력] 처리된 병합 파일

In [475]:
fold

'D:/data/big2/BD1_4등급_데이터_점검1/result_emis_update2'

In [476]:
# # about 33s
# df.to_csv(os.path.join(fold, f'4등급_차량_리스트_양식_result_처리_{today_date}.csv'), index=False, encoding='cp949')

In [477]:
submit_df = df[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자', '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명', '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명', '소유자구분', '말소여부', '일치유무', '배인번호_수정', '등급_수정', 'DPF유무_수정']]
submit_df = submit_df.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

In [478]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_PM_05.03.23'

In [479]:
# about 33s
submit_df.to_csv(os.path.join(submit_fold, f'제출_4등급_차량_리스트_양식_result_처리_{today_date}.csv'), index=False, encoding='cp949')

In [480]:
# about 17s
submit_df.iloc[:500000].to_csv(os.path.join(submit_fold, f'제출_4등급_차량_리스트_양식_result_처리1_{today_date}.csv'), index=False, encoding='cp949')
submit_df.iloc[500000:].to_csv(os.path.join(submit_fold, f'제출_4등급_차량_리스트_양식_result_처리2_{today_date}.csv'), index=False, encoding='cp949')

# 샘플 출력

## 조건1

### 3, 5 등급으로 등급 변경 차량
- 정상(norm)

In [182]:
norm.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

In [183]:
norm[(norm['배출가스인증번호'] == '8MY-KM-13-19') & ( (norm['등급_수정'] == "5") | (norm['등급_수정'] == "3") )].shape

(83, 69)

In [184]:
grade_chaneged = norm.loc[(norm['등급_수정'] == '3') | (norm['등급_수정'] == '5')]
grade_chaneged.shape

(4479, 69)

In [185]:
grade_chaneged.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

#### [출력] 3, 5 등급 변경 차량

In [186]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_AM_09.24.46'

In [187]:
# grade_chaneged.to_csv(os.path.join(total_norm_fold, f'3,5등급변경차량_{today_date}.csv'), index=False, encoding='cp949')

In [243]:
submit_grade_chaneged = grade_chaneged[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_grade_chaneged = submit_grade_chaneged.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

# about 33s
submit_grade_chaneged.to_csv(os.path.join(submit_fold, f'제출_3,5등급변경차량_{today_date}.csv'), index=False, encoding='cp949')

### 배인번호 변경된 차량
- 정상(norm)
- 미부여(nad) : nad1y, nad2y, nad3y, nad3n
- 제작사이상(weird) : weird1y, weird2y, weird2n

In [248]:
norm_ch = norm.loc[norm['배출가스인증번호'] != norm['배인번호_수정']]
norm_ch.shape

(127823, 69)

In [249]:
nad1y.shape, nad2y.shape, nad3y.shape, weird1y.shape, weird2y.shape

((6294, 69), (139, 69), (87, 69), (630, 69), (132, 69), (14, 69))

In [251]:
norm_ch.shape[0] + nad1y.shape[0] + nad2y.shape[0] + nad3y.shape[0] + weird1y.shape[0] + weird2y.shape[0]

134489

In [252]:
emis_ch = pd.concat([norm_ch, nad1y, nad2y, nad3y, weird1y, weird2y], ignore_index=True)
emis_ch.shape

(134489, 69)

In [253]:
emis_ch.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

#### [출력] 배인번호 변경된 차량

In [254]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_PM_12.43.21'

In [255]:
# emis_ch.to_csv(os.path.join(total_norm_fold, f'배인번호_변경된_차량_{today_date}.csv'), index=False, encoding='cp949')

In [256]:
submit_emis_ch = emis_ch[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_emis_ch = submit_emis_ch.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_emis_ch.to_csv(os.path.join(submit_fold, f'제출_배인번호_변경된_차량_{today_date}.csv'), index=False, encoding='cp949')

### 나머지 차량
- 비정상(error)
- 제작사이상(weird) : weird2n
- 미부여(nad) : nad2n

In [258]:
error.shape, weird2n.shape, nad3n.shape

((1309, 69), (4716, 69), (630, 69))

In [259]:
rest = pd.concat([error, weird2n, nad3n], ignore_index=True)
rest.shape

(6655, 69)

In [260]:
rest.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

#### [출력] 나머지 차량

In [261]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_PM_12.44.25'

In [262]:
# rest.to_csv(os.path.join(total_norm_fold, f'나머지_미부여_비정상_제작사이상_차량_{today_date}.csv'), index=False, encoding='cp949')

In [263]:
submit_rest = rest[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_rest = submit_rest.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_rest.to_csv(os.path.join(submit_fold, f'제출_나머지_미부여_비정상_제작사이상_차량_{today_date}.csv'), index=False, encoding='cp949')

### 최종 차량 수 확인

In [264]:
norm.loc[norm['배출가스인증번호'] == norm['배인번호_수정']].shape[0] + emis_ch.shape[0] + rest.shape[0]

1153813

## 조건2

### 제작사 확인 불가
- 1201 개

In [265]:
norm.loc[norm['배인번호_수정'] == '제작사확인불가'].shape

(1201, 69)

In [266]:
not_confirm_product = norm.loc[norm['배인번호_수정'] == '제작사확인불가']
not_confirm_product.columns

Index(['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '등급', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무', '사용연료', '주행거리', '차명_y', '차량번호_y', '차종', '차량연식',
       '엔진형식_y', '검사방법', '검사일자', '검사종류', '검사판정시각', '검사판정', '산소값', '이산화탄소값',
       '무부하매연판정1', '무부하매연판정2', '무부하매연판정3', '무부하매연판정4', '무부하매연판정5', '무부하매연판정6',
       '무부하매연허용치1', '무부하매연허용치2', '무부하매연허용치3', '무부하매연허용치4', '무부하매연허용치5',
       '무부하매연허용치6', '무부하매연측정치1', '무부하매연측정치2', '무부하매연측정치3', '무부하매연측정치4',
       '무부하매연측정치5', '무부하매연측정치6', '정기정밀', 'vin10', 'vin_year', 'DPF유무',
       '배인번호_수정', '등급_수정', 'DPF유무_수정', '확인유무', '비고'],
      dtype='object')

In [267]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_PM_12.44.43'

In [268]:
# norm.loc[norm['배인번호_수정'] == '제작사확인불가'].to_csv(os.path.join(total_norm_fold, f'제작사확인불가_{today_date}.csv'), index=False, encoding='cp949')

In [269]:
submit_not_confirm_product = not_confirm_product[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_not_confirm_product = submit_not_confirm_product.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_not_confirm_product.to_csv(os.path.join(submit_fold, f'제출_제작사확인불가_{today_date}.csv'), index=False, encoding='cp949')

### 배출가스 인증번호의 제작사 미매칭
- 4,862

In [270]:
mod_weird = pd.concat([weird1y, weird2y, weird2n], ignore_index=True)
mod_weird.shape

(4862, 69)

In [271]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_PM_12.44.53'

In [272]:
# mod_weird.to_csv(os.path.join(total_weird_fold, f'배출가스_인증번호의_제작사_미매칭_{today_date}.csv'), index=False, encoding='cp949')

In [273]:
submit_mod_weird = mod_weird[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_mod_weird = submit_mod_weird.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_mod_weird.to_csv(os.path.join(submit_fold, f'제출_배출가스_인증번호의_제작사_미매칭_{today_date}.csv'), index=False, encoding='cp949')

### 배출가스 인증번호 미기입 차량
- 7,150

In [274]:
mod_nad = pd.concat([nad1y, nad2y, nad3y, nad3n], ignore_index=True)
mod_nad.shape

(7150, 69)

In [275]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_PM_12.45.01'

In [276]:
# mod_nad.to_csv(os.path.join(total_nan_fold, f'미기입차량_수정_{today_date}.csv'), index=False, encoding='cp949')

In [277]:
submit_mod_nad = mod_nad[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_mod_nad = submit_mod_nad.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_mod_nad.to_csv(os.path.join(submit_fold, f'제출_미기입차량_수정_{today_date}.csv'), index=False, encoding='cp949')

## 조건3
- norm : 1140492
- error : 1309
- weird1y : 132
- weird1n
    - weird2y : 14
    - weird2n : 4716
- nad
    - nad1y : 6294
    - nad1n
        - nad2y : 139
        - nad2n
            - nad3y : 87
            - nad3n : 630

In [217]:
# 배인번호 정상 
    # total_norm = norm, weird1y, weird2y, nad1y, nad2y, nad3y : 1147158
# 배인번호 비정상
    # error : 1309
# 제작사 이상
    # weird2n : 4716
# 미부여
    # nad3n : 630

# 총 : 1153813

### 정상

In [218]:
total_norm = pd.concat([norm, weird1y, weird2y, nad1y, nad2y, nad3y], ignore_index=True)
total_norm.shape

(1147158, 69)

In [219]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

'2023.03.17_AM_09.30.14'

In [220]:
# total_norm.to_csv(os.path.join(total_norm_fold, 'total_norm.csv'), index=False, encoding='cp949')

In [237]:
submit_total_norm = total_norm[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_total_norm = submit_total_norm.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_total_norm.to_csv(os.path.join(submit_fold, f'total_norm_{today_date}.csv'), index=False, encoding='cp949')

### 비정상

In [222]:
error.shape

(1309, 69)

In [ ]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

In [236]:
submit_error = error[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_error = submit_error.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_error.to_csv(os.path.join(submit_fold, f'error_{today_date}.csv'), index=False, encoding='cp949')

### 제작사 이상

In [224]:
weird2n.shape

(4716, 69)

In [ ]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

In [235]:
submit_weird2n = weird2n[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_weird2n = submit_weird2n.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_weird2n.to_csv(os.path.join(submit_fold, f'weird2n_{today_date}.csv'), index=False, encoding='cp949')

### 미부여

In [226]:
nad3n.shape

(630, 69)

In [ ]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

In [234]:
submit_nad3n = nad3n[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자',
       '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명',
       '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명',
       '소유자구분', '말소여부', '일치유무']]
submit_nad3n = submit_nad3n.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_nad3n.to_csv(os.path.join(submit_fold, f'nad3n_{today_date}.csv'), index=False, encoding='cp949')

### 제작사 이상 변경사항 포함

In [228]:
mod_merge_weird = pd.concat([weird1y, weird2y, weird2n], ignore_index=True)
mod_merge_weird.shape

(4862, 69)

In [ ]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

In [244]:
submit_mod_merge_weird = mod_merge_weird[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자', '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명', '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명', '소유자구분', '말소여부', '일치유무']]
submit_mod_merge_weird = submit_mod_merge_weird.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_mod_merge_weird.to_csv(os.path.join(submit_fold, f'mod_merge_weird_{today_date}.csv'), index=False, encoding='cp949')

In [246]:
mod_merge_weird['배인번호_수정'].value_counts(dropna=True)

7MY-DK-14-5     124
5MY-HB-14-1       8
7MY-SM-14-36      2
7MY-HD-24-09      2
6MY-BK-13-08      2
7MY-KM-14-63      2
7MY-HD-14-46      1
6MY-KM-24-06      1
7MY-HD-24-29      1
7MY-MK-14-04      1
7MY-HD-24-28      1
FMY-VTK-14-1      1
Name: 배인번호_수정, dtype: int64

### 미부여 변경사항 포함

In [230]:
mod_merge_nad = pd.concat([nad1y, nad2y, nad3y, nad3n], ignore_index=True)
mod_merge_nad.shape

(7150, 69)

In [ ]:
today_date = datetime.today().strftime("%Y.%m.%d_%p_%I.%M.%S")
today_date

In [245]:
submit_mod_merge_nad = mod_merge_nad[['차대번호', '차량번호_x', '제원관리번호', '차종_원본', '용도', '최초등록일', '연식', '제작일자', '검사유효일', '배출가스인증번호', '배인번호_수정', '등급', '등급_수정', '차명_x', '차종분류', '차종유형', '자동차형식', '제작사명', '연료', '엔진형식_x', '총중량', '적재중량', '엔진출력', '배기량', '법정동코드', '시도명', '시군구명', '소유자구분', '말소여부', '일치유무']]
submit_mod_merge_nad = submit_mod_merge_nad.rename(columns={'차량번호_x':'차량번호', '차명_x':'차명', '엔진형식_x':'엔진형식'})

submit_mod_merge_nad.to_csv(os.path.join(submit_fold, f'mod_merge_nad_{today_date}.csv'), index=False, encoding='cp949')

In [247]:
mod_merge_nad['배인번호_수정'].value_counts(dropna=True)

5MY-HB-14-1      1669
6MY-DK-14-1      1001
6MY-PA-14-4       965
8MY-HB-14-1       764
7MY-DK-14-5       437
6MY-DK-14-2       260
8MY-DK-14-01      249
8MY-PA-14-05      204
5MY-HB-14-4       127
5MY-PA-14-17      121
5MY-PA-14-20       89
7MY-PA-14-7        85
8MY-PA-14-04       80
8MY-DK-14-02       77
7MY-DK-14-4        65
8MY-FK-14-1        53
8MY-PA-14-03       50
7MY-PA-13-09       48
8MY-FK-14-2        47
8MY-DK-14-06       40
7MY-PA-14-2        21
FMY-VTK-14-1       12
5MY-BK-14-07        9
3MY-HD-14           8
6MY-KM-24-81        6
4MY-KM-07           4
5MY-KM-24-32        4
BMY-HD-14-54        3
7MY-HD-24-30        2
7MY-DK-14-05        1
7MY-PA-13-11        1
BMY-KM-14-32        1
FMY-VTK-14-2        1
4MY-KM-06           1
7MY-PA-14-1         1
7MY-PA-13-08        1
7MY-KM-14-63        1
4MY-KM-03           1
7MY-VTK-24-01       1
8MY-KM-13-19        1
7MY-DW-13-07        1
6MY-HD-24-44        1
7MY-SM-14-36        1
6MY-HD-14-63        1
5MY-PA-14-18        1
4MY-AD-20 

## 정상 배인번호별 출력

In [278]:
total_norm.shape

(1147158, 69)

In [288]:
total_norm.groupby(['배출가스인증번호', '배인번호_수정'], dropna=False)['차대번호'].count()

배출가스인증번호      배인번호_수정     
4MY-AD-20     4MY-AD-20        29
4MY-DC-33     6MY-DC-23-17      7
4MY-PA-06     제작사확인불가          13
5MY-AD-13-06  5MY-AD-13-06    348
5MY-AD-13-07  5MY-AD-13-07    118
                             ... 
NaN           8MY-PA-14-05    204
              BMY-HD-14-54      3
              BMY-KM-14-32      1
              FMY-VTK-14-1     12
              FMY-VTK-14-2      1
Name: 차대번호, Length: 223, dtype: int64

In [289]:
total_norm.groupby(['배출가스인증번호', '배인번호_수정'], dropna=False)['차대번호'].count().to_excel(os.path.join(total_norm_fold, '통계_total_norm.xlsx'))

# 분석

## 3, 5 등급 변경 차량

In [481]:
submit_df['등급_수정'] = submit_df['등급_수정'].astype('str')

In [482]:
submit_df.loc[submit_df['등급_수정'] == '3'].shape

(4378, 32)

In [483]:
submit_df.loc[submit_df['등급_수정'] == '5'].shape

(124, 32)

# CODE End